### PR TITLE
[Android] Rename System.Security.Cryptography.Native.OpenSsl on Android

### LINKED ISSUE #52327: Rename libSystem.Security.Cryptography.Native.OpenSsl on Android
As part of the cleanup work in https://github.com/dotnet/runtime/issues/45741, we should rename the crypto library on Android since it does not have anything to do w/ OpenSSL. 

Name of library: System.Security.Cryptography.Native.Android

You'll know it's right by checking the file in $(RepoRoot)/artifacts/bin/mono/<rid>.<Configuration>/ is named libSystem.Security.Cryptography.Native.Android.a

_Note on osx / linux, we prepend lib to the file name._  

The rename starts at https://github.com/dotnet/runtime/blob/f64246ce08fb7a58221b2b7c8e68f69c02522b0d/src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/CMakeLists.txt#L45-L47 and will likely need to be modified elsewhere.  

Another important place is here https://github.com/dotnet/runtime/blob/c9c954be788c4d4d13af326d83ba206a7f2ce45d/src/tasks/AndroidAppBuilder/Templates/MonoRunner.java#L34

### PR DESCRIPTION
**Pull request by tqiu8**
Rename `System.Security.Cryptography.Native.OpenSsl` to `System.Security.Cryptography.Native.Android`
Fix #52327

### CODE REVIEW

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props
Comment (filipnavara) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:51]: This is not Android specific, should stay unchanged.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props
Comment (filipnavara) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:96]: This is not Android specific, should stay unchanged.

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (filipnavara) [inline src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs:11]: I'm pretty sure this is used on non-Android platforms.

File: src/libraries/Common/tests/TestUtilities/Interop/Interop.Libraries.cs
Comment (filipnavara) [inline src/libraries/Common/tests/TestUtilities/Interop/Interop.Libraries.cs:9]: This is not Android specific...

File: src/native/corehost/hostpolicy/hostpolicy_context.cpp
Comment (filipnavara) [inline src/native/corehost/hostpolicy/hostpolicy_context.cpp:78]: This is not Android code.

File: src/native/corehost/apphost/static/CMakeLists.txt
Comment (filipnavara) [inline src/native/corehost/apphost/static/CMakeLists.txt:144]: This is not Android code.

File: src/native/corehost/apphost/static/CMakeLists.txt
Comment (filipnavara) [inline src/native/corehost/apphost/static/CMakeLists.txt:164]: This is not Android code.

**Review by filipnavara — COMMENTED** [review]
Most of the places should actually remain unchanged since they reference the real OpenSSL shim library and not the Android one.

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (filipnavara) [reply src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs:11]: There are different possible ways to deal with this:
- Split the Interop.Libraries.cs in a way that on Android builds you get a different name. 
- Add a different name in https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/Interop/Android/Interop.Libraries.cs and then update the references in the Interop files to use it (eg.
https://github.com/dotnet/runtime/blob/66d26c36991c7b5ff09c7af07cae9c6d59510b7c/src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.Bignum.cs#L13)

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props
Comment (tqiu8) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:51]: Ah got it. I wasn't sure on the extent of renaming. Thank you for helping me catch all the non-Android specific ones.

File: src/libraries/pretest.proj
Comment (akoeplinger) [inline src/libraries/pretest.proj:41]: the .Android.so should be added in addition instead of replacing, since .OpenSsl.so still exists for e.g. desktop Linux.

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/CMakeLists.txt
Comment (akoeplinger) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/CMakeLists.txt:47]: this TODO comment can be removed now

**Review by akoeplinger — COMMENTED** [review]
Looks good apart from two comments

**Comment by akoeplinger** [comment]
There are quite a lot of test failures with `System.DllNotFoundException : libSystem.Security.Cryptography.Native.OpenSsl` on the runtime-staging Android legs, e.g.

```
   at System.PlatformDetection.GetAlpnSupport() in /_/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:line 162
   at System.Lazy`1[[System.Boolean, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ViaFactory(LazyThreadSafetyMode mode) in System.Private.CoreLib.dll:token 0x6000fc6+0x43
   at System.Lazy`1[[System.Boolean, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor) in System.Private.CoreLib.dll:token 0x6000fc7+0x22
   at System.Lazy`1[[System.Boolean, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].CreateValue() in System.Private.CoreLib.dll:token 0x6000fcc+0x74
   at System.Lazy`1[[System.Boolean, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].get_Value() in System.Private.CoreLib.dll:token 0x6000fd2+0xa
   at System.PlatformDetection.get_SupportsAlpn() in /_/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:line 168
   at System.Net.Test.Common.GenericLoopbackOptions..ctor() in /_/src/libraries/Common/tests/System/Net/Http/GenericLoopbackServer.cs:line 96
   at System.Net.Test.Common.LoopbackServer.Options..ctor() in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 384
   at System.Net.Test.Common.LoopbackServer..ctor(Options options) in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 29
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in /_/src/libraries/Common/tests/System/Net/Http/LoopbackServer.cs:line 70
   at System.Net.Tests.WebClientTestBase.DownloadFile_Success() in /_/src/libraries/System.Net.WebClient/tests/WebClientTest.cs:line 583
--- End of stack trace from previous location ---
```

This is because we include the Unix/OpenSSL version of Interop.CryptoNative.cs which has the wrong library name.

I think we should rename the `CryptoNative` field to `AndroidCryptoNative` for all the DllImports to `AndroidCryptoNative_` functions to make these mistakes easier to catch.

**Comment by tqiu8** [comment]
> This is because we include the Unix/OpenSSL version of Interop.CryptoNative.cs which has the wrong library name.
> 
> I think we should rename the `CryptoNative` field to `AndroidCryptoNative` for all the DllImports to `AndroidCryptoNative_` functions to make these mistakes easier to catch.

That was one of my original approaches, but Android still uses several Interop files in the src/libraries/Common/src/Interop/Unix directory (Dllimports into `CryptoNative_`), which would require splitting those files into the src/libraries/Common/src/Interop/Android directory as well, thus creating duplicate files. Renaming `CryptoNative` is more efficient in that it takes advantage of shared entrypoint names but is definitely more ambiguous and harder to catch. I suppose in the long run it's better to make the split between Android and Unix more distinct? 

/cc @steveisok

**Comment by marek-safar** [comment]
Can this be merged now?

**Comment by filipnavara** [comment]
> Can this be merged now?

I believe it is still broken and references a wrong name of the library in some of the managed code (per comments above and Helix results in the runtime-staging pipelines).

**Comment by tqiu8** [comment]
From what I can tell, test failures are seemingly unrelated, re-running some of the failing lanes in case some of them were flakey.

**Comment by filipnavara** [comment]
FWIW it looks good to me. It would likely make sense to rename `OpenSslCryptographicException` to something Android specific but I don't think it's worth to hold up the PR because of it since it's a small detail and this PR would unblock any potential fixes of the name on the Xamarin side.

**Comment by steveisok** [comment]
@tqiu8 It looks like you're hitting known flakes.  The android arm64 failures are being tracked https://github.com/dotnet/xharness/issues/597 and the Android x86 failure is hitting https://github.com/dotnet/runtime/issues/52649

**Review by steveisok — APPROVED** [review]

**Pull request merged**
